### PR TITLE
[BEAM-115] Translate pipeline graph to and from Runner API protos.

### PR DESCRIPTION
**Pull request by robertwb**
There are some caveates:
  * Specific known transforms, with their payloads, are not yet translated.
  * Side inputs are not yet supported.

All pipelines without side inputs are passed through this translation by default
before execution.

Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by robertwb** [comment]
R: @vikkyrk

File: sdks/python/apache_beam/pipeline.py [hunk 1]
```diff
@@ -441,3 +490,47 @@ def visit(self, visitor, pipeline, visited):
         if v not in visited:
           visited.add(v)
           visitor.visit_value(v, self)
+
+  def named_inputs(self):
+    # TODO(robertwb): Push names up into the sdk construction.
```
Comment (vikkyrk) [inline sdks/python/apache_beam/pipeline.py]: Maybe track these TODOs in an umbrella jira?

File: sdks/python/apache_beam/pipeline_test.py [hunk 2]
```diff
@@ -439,6 +440,21 @@ def test_dir(self):
              if not attr.startswith('_')]))
 
 
+class RunnerApiTest(unittest.TestCase):
+
+  def test_simple(self):
+    """Tests serializing, deserializing, and running a simple pipeline.
+
+    More extensive tests are done at pipeline.run for each suitable test.
```
Comment (vikkyrk) [inline sdks/python/apache_beam/pipeline_test.py]: Are these pipeline.run tests actually enabled for runner api?

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 3]
```diff
@@ -219,17 +219,17 @@ def _get_side_input_encoding(self, input_encoding):
   def _get_encoded_output_coder(self, transform_node, window_value=True):
     """Returns the cloud encoding of the coder for the output of a transform."""
     if (len(transform_node.outputs) == 1
-        and transform_node.outputs[0].element_type is not None):
+        and transform_node.outputs[None].element_type is not None):
```
Comment (vikkyrk) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: For my understanding, why change this from 0 to `None`?

File: sdks/python/apache_beam/pipeline.py [hunk 4]
```diff
@@ -299,6 +308,42 @@ def apply(self, transform, pvalueish=None, label=None):
     self.transforms_stack.pop()
     return pvalueish_result
 
+  def _verify_runner_api_compatible(self):
+    class Visitor(PipelineVisitor):  # pylint: disable=used-before-assignment
+      ok = True  # Really a nonlocal.
```
Comment (robertwb) [inline sdks/python/apache_beam/pipeline.py]: It's an attribute (adding a getter method would probably be overkill). The comment is just that I'd use a non-local if it were available in Python 2.

File: sdks/python/apache_beam/examples/snippets/snippets.py [hunk 5]
```diff
@@ -683,7 +683,7 @@ def expand(self, pcoll):
       lines, beam.equal_to(
           ['line ' + str(number) for number in range(0, count)]))
 
-  p.run().wait_until_finish()
+  p.run(test_runner_api=False).wait_until_finish()  # Pickle errors
```
Comment (robertwb) [inline sdks/python/apache_beam/examples/snippets/snippets.py]: Done.

File: sdks/python/apache_beam/pipeline_test.py (hunk 2 above)
Comment (robertwb) [inline sdks/python/apache_beam/pipeline_test.py]: Yes, pipeline.run has been modified to do a round trip through protos for all possible pipelines.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py (hunk 3 above)
Comment (robertwb) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: This goes with the change in Pipeline.add_output, to be more consistent.

File: sdks/python/apache_beam/pipeline.py (hunk 1 above)
Comment (robertwb) [inline sdks/python/apache_beam/pipeline.py]: Done.

**Comment by robertwb** [comment]
Thanks, PTAL.

File: sdks/python/apache_beam/pipeline_test.py (hunk 2 above)
Comment (vikkyrk) [inline sdks/python/apache_beam/pipeline_test.py]: Ah I missed that it was set to `True` by defaults. Sounds good.

**Comment by sb2nov** [comment]
Random question: Why did we change the tag for the main output to None as that is a breaking change for the runners.

**Pull request closed without merging**